src/gui/properties/piecesbar.cpp

```diff
@@ -0,0 +1,318 @@
+        return true;
+    }
+    else {
+        return base::event(e);
+    }
+}
+
+void PiecesBar::enterEvent(QEvent *e)
+{
+    m_hovered = true;
+    base::enterEvent(e);
+}
+
+void PiecesBar::leaveEvent(QEvent *e)
+{
+    m_hovered = false;
+    m_highlitedRegion = QRect();
+    requestImageUpdate();
+    base::leaveEvent(e);
+}
+
+void PiecesBar::mouseMoveEvent(QMouseEvent *e)
+{
+    // if user pointed to a piece which is a part of a single large file,
+    // we highlight the space, occupied by this file
+    highlightFile(e->pos().x() - borderWidth);
+    base::mouseMoveEvent(e);
+}
+
+void PiecesBar::paintEvent(QPaintEvent*)
+{
+    QPainter painter(this);
+    QRect imageRect(borderWidth, borderWidth, width() - 2 * borderWidth, height() - 2 * borderWidth);
+    if (m_image.isNull()) {
+        painter.setBrush(Qt::white);
+        painter.drawRect(imageRect);
+    }
+    else {
+        if (m_image.width() != imageRect.width())
+            updateImage(m_image);
+        painter.drawImage(imageRect, m_image);
+    }
+
+    if (!m_highlitedRegion.isNull()) {
+        QColor highlightColor {this->palette().color(QPalette::Active, QPalette::Highlight)};
+        highlightColor.setAlphaF(0.35);
+        QRect targetHighlightRect {m_highlitedRegion.adjusted(borderWidth, borderWidth, borderWidth, height() - 2 * borderWidth)};
+        painter.fillRect(targetHighlightRect, highlightColor);
+    }
+
+    QPainterPath border;
+    border.addRect(0, 0, width(), height());
+    painter.setPen(m_borderColor);
+    painter.drawPath(border);
+}
+
+void PiecesBar::requestImageUpdate()
+{
+    if (updateImage(m_image))
+        update();
+}
+
+QColor PiecesBar::backgroundColor() const
+{
+    return m_bgColor;
+}
+
+QColor PiecesBar::borderColor() const
+{
+    return m_borderColor;
+}
+
+QColor PiecesBar::pieceColor() const
+{
+    return m_pieceColor;
+}
+
+const QVector<QRgb> &PiecesBar::pieceColors() const
+{
+    return m_pieceColors;
+}
+
+QRgb PiecesBar::mixTwoColors(QRgb rgb1, QRgb rgb2, float ratio)
+{
+    int r1 = qRed(rgb1);
+    int g1 = qGreen(rgb1);
+    int b1 = qBlue(rgb1);
+
+    int r2 = qRed(rgb2);
+    int g2 = qGreen(rgb2);
+    int b2 = qBlue(rgb2);
+
+    float ratioN = 1.0f - ratio;
+    int r = (r1 * ratioN) + (r2 * ratio);
+    int g = (g1 * ratioN) + (g2 * ratio);
+    int b = (b1 * ratioN) + (b2 * ratio);
+
+    return qRgb(r, g, b);
+}
+
+void PiecesBar::showToolTip(const QHelpEvent *e)
+{
+    if (!m_torrent)
+        return;
+
+    QString toolTipText;
+    QTextStream stream(&toolTipText, QIODevice::WriteOnly);
+    bool showDetailedInformation = QApplication::keyboardModifiers().testFlag(Qt::ShiftModifier);
+    if (showDetailedInformation) {
+        stream << "<html><body>";
+        const int imagePos = e->pos().x() - borderWidth;
+        if ((imagePos >=0) && (imagePos < m_image.width())) {
+            PieceIndexToImagePos transform {m_torrent->info(), m_image};
```

Comment:
But we did accept using it!
Let's be consistent in our actions.

@evsh You agreed to this so easily? After you have fought for this opportunity? You surprise me (not to say otherwise).